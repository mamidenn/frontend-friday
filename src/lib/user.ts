/**
 * # Including/Excluding fields
 *
 * Sometimes, maybe you want some random names and not extraneous data such as location, phone, etc.
 * Using the inc and exc parameters, you can specify which fields to include or exclude
 * respectively.
 *
 * By specifying only the fields you want, the generator can save time by skipping CPU intensive
 * fields like "login" for example. These parameters accept the following values in a comma
 * delimited list
 *
 * - gender
 * - name
 * - location
 * - email
 * - login
 * - registered
 * - dob
 * - phone
 * - cell
 * - id
 * - picture
 * - nat
 *
 * If you only wanted the names, genders, and nats of users:
 *
 * ```txt
 * https://randomuser.me/api/?inc=gender,name,nat
 * ```
 */
export const getUser = async function (...fields) {
	// please do not change the seed, the tests depend on it
	const response = await fetch('https://randomuser.me/api/?seed=frontendfriday');
	const json: UserApiResult = await response.json();
	return json.results[0];
};

export interface UserApiResult {
	results: User[];
	info: Info;
}

export interface Info {
	seed: string;
	results: number;
	page: number;
	version: string;
}

export interface User {
	gender: string;
	name: Name;
	location: Location;
	email: string;
	login: Login;
	dob: DateAndAge;
	registered: DateAndAge;
	phone: string;
	cell: string;
	id: ID;
	picture: Picture;
	nat: string;
}

export interface DateAndAge {
	date: Date;
	age: number;
}

export interface ID {
	name: string;
	value: string;
}

export interface Location {
	street: Street;
	city: string;
	state: string;
	country: string;
	postcode: number;
	coordinates: Coordinates;
	timezone: Timezone;
}

export interface Coordinates {
	latitude: string;
	longitude: string;
}

export interface Street {
	number: number;
	name: string;
}

export interface Timezone {
	offset: string;
	description: string;
}

export interface Login {
	uuid: string;
	username: string;
	password: string;
	salt: string;
	md5: string;
	sha1: string;
	sha256: string;
}

export interface Name {
	title: string;
	first: string;
	last: string;
}

export interface Picture {
	large: string;
	medium: string;
	thumbnail: string;
}
