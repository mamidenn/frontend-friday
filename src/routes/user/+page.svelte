<script lang="ts">
	import type { PageData } from './$types';

	export let data: PageData;
	$: ({ user } = data);
</script>

<svelte:head>
	<title>User</title>
</svelte:head>

<img
	src={user.picture.large}
	alt="A picture of {user.name.title} {user.name.first} {user.name.last}"
/>

<h2>{user.name.title} {user.name.first} {user.name.last}</h2>

<table>
	<tr>
		<th>Location</th>
		<td>{user.location.city}, {user.location.state}, {user.location.country}</td>
	</tr>
	<tr>
		<th>Email</th>
		<td>{user.email}</td>
	</tr>
	<tr>
		<th>Phone</th>
		<td>{user.phone}</td>
	</tr>
	<tr>
		<th>Mobile</th>
		<td>{user.cell}</td>
	</tr>
	<tr>
		<th>Date Of Birth</th>
		<td>{user.dob.date.toDateString()} ({user.dob.age})</td>
	</tr>
</table>

<style lang="postcss">
</style>
