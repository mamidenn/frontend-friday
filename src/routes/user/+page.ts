import { getUser } from '$lib/user';
import type { PageLoad } from './$types';

export const load = (async ({ fetch }) => {
	return { user: await getUser(fetch) };
}) satisfies PageLoad;
