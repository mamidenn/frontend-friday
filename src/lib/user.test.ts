import { describe, expect, expectTypeOf, it } from 'vitest';
import { getUser, type ID, type Name, type User } from './user';

describe('getUser', () => {
	it('returns a single random user', async () => {
		const user = await getUser();

		expect(user).toMatchSnapshot();
		// red squiggles on this line mean test failure
		expectTypeOf(user).toEqualTypeOf<User>();
	});

	describe('given it is passed some fields', () => {
		it('only returns the requested fields (name)', async ({ expect }) => {
			const user = await getUser('name');

			expect(user).toMatchInlineSnapshot(`
				{
				  "name": {
				    "first": "Gaspard",
				    "last": "Jean",
				    "title": "Mr",
				  },
				}
			`);
			// red squiggles on this line mean test failure
			expectTypeOf(user).toEqualTypeOf<{ name: Name }>();
		});

		it('only returns the requested fields (id,name)', async ({ expect }) => {
			const user = await getUser('id', 'name');

			expect(user).toMatchInlineSnapshot(`
				{
				  "id": {
				    "name": "INSEE",
				    "value": "1570185270955 82",
				  },
				  "name": {
				    "first": "Gaspard",
				    "last": "Jean",
				    "title": "Mr",
				  },
				}
			`);
			// red squiggles on this line mean test failure
			expectTypeOf(user).toEqualTypeOf<{ id: ID; name: Name }>();
		});

		it('only returns the requested fields (id,name,email)', async ({ expect }) => {
			const user = await getUser('id', 'name', 'email');

			expect(user).toMatchInlineSnapshot(`
				{
				  "email": "gaspard.jean@example.com",
				  "id": {
				    "name": "INSEE",
				    "value": "1570185270955 82",
				  },
				  "name": {
				    "first": "Gaspard",
				    "last": "Jean",
				    "title": "Mr",
				  },
				}
			`);
			// red squiggles on this line mean test failure
			expectTypeOf(user).toEqualTypeOf<{ id: ID; name: Name; email: string }>();
		});
	});
});
