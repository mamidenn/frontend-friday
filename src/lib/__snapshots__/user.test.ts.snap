// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getUser > returns a single random user 1`] = `
{
  "cell": "06-96-21-80-09",
  "dob": {
    "age": 66,
    "date": 1957-02-10T12:12:14.050Z,
  },
  "email": "gaspard.jean@example.com",
  "gender": "male",
  "id": {
    "name": "INSEE",
    "value": "1570185270955 82",
  },
  "location": {
    "city": "Nîmes",
    "coordinates": {
      "latitude": "-17.3178",
      "longitude": "-49.9042",
    },
    "country": "France",
    "postcode": 74133,
    "state": "Ariège",
    "street": {
      "name": "Rue de L'Abbé-Groult",
      "number": 7210,
    },
    "timezone": {
      "description": "Pacific Time (US & Canada)",
      "offset": "-8:00",
    },
  },
  "login": {
    "md5": "8f7ee1c713065e5e6f076bd0cacf65e6",
    "password": "crusher",
    "salt": "OmkZGgA8",
    "sha1": "94b43e8dc8dbf369affb967b15b699a9131d6500",
    "sha256": "56d5a11758d681371b3a91bf3f8f60a38d48c731b71af19198ed0b687eeaccfd",
    "username": "heavysnake221",
    "uuid": "572b8a1b-a32d-4816-b47d-23bd3449ddb8",
  },
  "name": {
    "first": "Gaspard",
    "last": "Jean",
    "title": "Mr",
  },
  "nat": "FR",
  "phone": "02-32-58-87-55",
  "picture": {
    "large": "https://randomuser.me/api/portraits/men/86.jpg",
    "medium": "https://randomuser.me/api/portraits/med/men/86.jpg",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/86.jpg",
  },
  "registered": {
    "age": 7,
    "date": 2016-07-10T13:27:18.462Z,
  },
}
`;
